File: BTFViewer/btf_viewer_pkg/loading_state.py
"""Loading-state UX (Step 3).

Lockstep with ``web/src/utils/loadingState.js``.
"""
from __future__ import annotations

import re
from typing import Dict, Tuple

LOADING_STAGES: Dict[str, str] = {
    "reading": "Reading trace…",
    "parsing": "Parsing events…",
    "building": "Building Timeline…",
    "computing": "Computing Statistics…",
    "opening": "Opening trace…",
    "restoring": "Restoring session…",
    "demo": "Loading demo trace…",
}

_INTERNAL_STAGE_RULES: Tuple[Tuple[re.Pattern[str], str], ...] = (
    (re.compile(r"reading file", re.I), "reading"),
    (re.compile(r"restoring session", re.I), "restoring"),
    (re.compile(r"demo trace", re.I), "demo"),
    (re.compile(r"packing trace|opening trace", re.I), "opening"),
    (re.compile(r"preparing statistics|computing statistics", re.I), "computing"),
    (
        re.compile(
            r"building scene|building legend|building task lod|building core lod|"
            r"per-task core lod|finalising segment|building sti",
            re.I,
        ),
        "building",
    ),
    (
        re.compile(
            r"reconstruct|lookup|index|sort|pair|tag channel|finalis|cul|analys|"
            r"sti channel|tick health",
            re.I,
        ),
        "parsing",
    ),
)


def resolve_loading_stage(internal_msg: str = "") -> str:
    msg = str(internal_msg or "").strip()
    if not msg:
        return "reading"
    for pattern, stage in _INTERNAL_STAGE_RULES:
        if pattern.search(msg):
            return stage
    return "parsing"


def format_loading_message(internal_msg: str = "") -> str:
    stage = resolve_loading_stage(internal_msg)
    return LOADING_STAGES.get(stage, LOADING_STAGES["parsing"])


def format_loading_pct(pct: float) -> str:
    try:
        n = float(pct)
    except (TypeError, ValueError):
        return ""
    if not n or n <= 0:
        return ""
    if n >= 100:
        return "100"
    rounded = round(n / 5) * 5
    return str(max(5, min(99, int(rounded))))


def is_loading_cancellable(phase: str = "parse") -> bool:
    return phase in ("parse", "read")


def normalize_loading_progress(
    pct: float, internal_msg: str = "", phase: str = "parse"
) -> dict:
    return {
        "pct": float(pct or 0),
        "msg": format_loading_message(internal_msg),
        "pct_label": format_loading_pct(pct),
        "cancellable": is_loading_cancellable(phase),
    }

File: BTFViewer/web/src/utils/loadingState.js
/**
 * Loading-state UX (Step 3).
 * Lockstep with btf_viewer_pkg/loading_state.py.
 *
 * Maps implementation-oriented parser progress to user-facing stages and
 * avoids fake percentage precision in the load overlay.
 */

export const LOADING_STAGES = Object.freeze({
  reading: 'Reading trace…',
  parsing: 'Parsing events…',
  building: 'Building Timeline…',
  computing: 'Computing Statistics…',
  opening: 'Opening trace…',
  restoring: 'Restoring session…',
  demo: 'Loading demo trace…',
})

/** Internal parser / worker message patterns → stage key. Order matters. */
const INTERNAL_STAGE_RULES = [
  [/reading file/i, 'reading'],
  [/restoring session/i, 'restoring'],
  [/demo trace/i, 'demo'],
  [/packing trace|opening trace/i, 'opening'],
  [/preparing statistics|computing statistics/i, 'computing'],
  [/building scene|building legend|building task lod|building core lod|per-task core lod|finalising segment|building sti/i, 'building'],
  [/reconstruct|lookup|index|sort|pair|tag channel|finalis|cul|analys|sti channel|tick health/i, 'parsing'],
]

/**
 * Resolve a canonical loading stage key from an internal progress message.
 * @param {string} [internalMsg]
 * @returns {keyof typeof LOADING_STAGES}
 */
export function resolveLoadingStage(internalMsg) {
  const msg = String(internalMsg || '').trim()
  if (!msg) return 'reading'
  for (const [re, stage] of INTERNAL_STAGE_RULES) {
    if (re.test(msg)) return stage
  }
  return 'parsing'
}

/**
 * User-facing stage label for the load overlay / toolbar badge.
 * @param {string} [internalMsg]
 * @returns {string}
 */
export function formatLoadingMessage(internalMsg) {
  const stage = resolveLoadingStage(internalMsg)
  return LOADING_STAGES[stage] || LOADING_STAGES.parsing
}

/**
 * Display percentage — rounded to 5% steps; empty when unknown/zero.
 * @param {number} pct
 * @returns {string}
 */
export function formatLoadingPct(pct) {
  const n = Number(pct)
  if (!Number.isFinite(n) || n <= 0) return ''
  if (n >= 100) return '100'
  const rounded = Math.round(n / 5) * 5
  return String(Math.max(5, Math.min(99, rounded)))
}

/** Whether the current load phase supports user cancellation. */
export function isLoadingCancellable(phase) {
  return phase === 'parse' || phase === 'read'
}

/**
 * Normalize a progress callback payload for UI binding.
 * @param {number} pct
 * @param {string} [internalMsg]
 * @returns {{ pct: number, msg: string, pctLabel: string, cancellable: boolean }}
 */
export function normalizeLoadingProgress(pct, internalMsg, phase = 'parse') {
  return {
    pct: Number(pct) || 0,
    msg: formatLoadingMessage(internalMsg),
    pctLabel: formatLoadingPct(pct),
    cancellable: isLoadingCancellable(phase),
  }
}
